_posts/2020-10-19-torch-image-classification/torch-image-classification.Rmd

```diff
@@ -0,0 +1,543 @@
+     - data/bird_species/test/'ALEXANDRINE PARAKEET'/2.jpg
+     - data/bird_species/test/'ALEXANDRINE PARAKEET'/3.jpg
+     - data/bird_species/test/'ALEXANDRINE PARAKEET'/4.jpg
+     - data/bird_species/test/'ALEXANDRINE PARAKEET'/5.jpg
+     
+     data/test/'AMERICAN BITTERN'/
+     - data/bird_species/test/'AMERICAN BITTERN'/1.jpg
+     - data/bird_species/test/'AMERICAN BITTERN'/2.jpg
+     - data/bird_species/test/'AMERICAN BITTERN'/3.jpg
+     - data/bird_species/test/'AMERICAN BITTERN'/4.jpg
+     - data/bird_species/test/'AMERICAN BITTERN'/5.jpg
+
+This is exactly the kind of layout expected by `torch`s `image_folder_dataset()` -- and really `bird_species_dataset()` instantiates a subtype of this class. Had we downloaded the data manually, respecting the required directory structure, we could have created the datasets like so:
+
+```{r}
+# e.g.
+train_ds <- image_folder_dataset(
+  file.path(data_dir, "train"),
+  transform = train_transforms)
+```
+
+Now that we got the data, let us see how many items there are in each set.
+
+```{r}
+train_ds$.length()
+valid_ds$.length()
+test_ds$.length()
+```
+
+    31316
+    1125
+    1125
+
+That training set is really big! It's thus recommended to run this on GPU, or just play around with the provided Colab notebook.
+
+With so many samples, we're curious how many classes there are.
+
+```{r}
+class_names <- test_ds$classes
+length(class_names)
+```
+
+    225
+
+So we *do* have a substantial training set, but the task is formidable as well: We're going to tell apart no less than 225 different bird species.
+
+#### Data loaders
+
+While *datasets* know what to do with each single item, *data loaders* know how to treat them collectively. How many samples make up a batch? Do we want to feed them in the same order always, or instead, have a different order chosen for every epoch?
+
+```{r}
+batch_size <- 64
+
+train_dl <- dataloader(train_ds, batch_size = batch_size, shuffle = TRUE)
+valid_dl <- dataloader(valid_ds, batch_size = batch_size)
+test_dl <- dataloader(test_ds, batch_size = batch_size)
+```
+
+Data loaders, too, may be queried for their length. Now length means: How many batches?
+
+```{r}
+train_dl$.length() 
+valid_dl$.length() 
+test_dl$.length()  
+```
+
+    490
+    18
+    18
+
+#### Some birds
+
+Next, let's view a few images from the test set. We can retrieve the first batch -- images and corresponding classes -- by creating an iterator from the `dataloader` and calling `next()` on it:
+
+```{r}
+# for display purposes, here we are actually using a batch_size of 24
+batch <- train_dl$.iter()$.next()
+```
+
+`batch` is a list, the first item being the image tensors:
+
+```{r}
+batch[[1]]$size()
+```
+
+    [1]  24   3 224 224
+
+And the second, the classes:
+
+```{r}
+batch[[2]]$size()
+```
+
+    [1] 24
+
+Classes are coded as integers, to be used as indexes into a vector of class names. We'll use those for labeling the images.
```

Comment:
```suggestion
Classes are coded as integers, to be used as indices in a vector of class names. We'll use those for labeling the images.
```

Reply:
lol :-)
it's true that I was a DB admin for some years, but I spent even more years learning Latin ;-)))